### PR TITLE
Redirect test output to file, if this setting is enabled.

### PR DESCRIPTION
**Pull request by BogdanStirbat**
Issue: #986

## Overview

Hi!

I inspected the way test output are redirected to file, in surefire-junit4 provider. 
The main surefire-junit4 provider class can be found [here](https://github.com/apache/maven-surefire/blob/master/surefire-providers/surefire-junit4/src/main/java/org/apache/maven/surefire/junit4/JUnit4Provider.java).

I've noticed similarities, between surefire-junit4 and junit5 provider. 
When maven instantiates the surefire provider, it sends an object of type `ProviderParameters` to the surefire provider (in main class's constructor). This object is used to create an `ReporterFactory`, then an `RunListener`. This `RunListener` is actually an interface. Running the debugger, I saw that the actual implementation is an object of type `ForkingRunListener`. `ForkingRunListener` also implements `ConsoleOutputReceiver`. 

Method `ConsoleOutputCapture.startCapture(ConsoleOutputReceiver target)` is responsible for the actual output redirection. This method is invoked in surefire-junit4, but not in junit5 provider.

After this change, the tests output are redirected correctly.

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [ ] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by marcphilipp** [comment]
Can we write an automated test for capturing the output? How did you test it?

**Comment by BogdanStirbat** [comment]
Hi,

I don't know, how we can write an automated test. 
As far as I know, maven itself is responsible for the actual output redirection. I don't know any method, or if such method exists, to capture the plugin's output file.

Regarding testing, I followed following steps:

1. I created a dummy Java project, built with maven. 
2. I built the junit5 project (clean, with no modification after sources were downloaded), and I published it to the local repository. 
3. I configured the dummy Java project, created at 1, to: use junit5 surefire provider, having version I recently published to my local maven repository, and to redirect test output to file. I rebuilt the project, and I reproduced the bug. 
4. I modified the junit5 surefire provider (with modifications from this pull request), I rebuild and republished the junit5 provider to my local maven repository. 
5. I built again the dummy Java project. This time, I saw the tests output were no longer redirected to console.

File: junit-platform-surefire-provider/src/test/java/org/junit/platform/surefire/provider/JUnitPlatformProviderTests.java [hunk 1]
```diff
@@ -234,7 +235,7 @@ private static ProviderParameters providerParametersMock(Class<?>... testClasses
 		when(runOrderCalculator.orderTestClasses(any())).thenReturn(testsToRun);
 
 		ReporterFactory reporterFactory = mock(ReporterFactory.class);
-		RunListener runListener = mock(RunListener.class);
+		RunListener runListener = mock(ForkingRunListener.class);
```
[the remark this answers is not in the document]
Comment (BogdanStirbat) [reply junit-platform-surefire-provider/src/test/java/org/junit/platform/surefire/provider/JUnitPlatformProviderTests.java]: Hi,

Thanks for the observation. 
I've added the change.

File: junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/JUnitPlatformProvider.java [hunk 2]
```diff
@@ -120,6 +122,7 @@ private RunResult invokeAllTests(TestsToRun testsToRun) {
 		ReporterFactory reporterFactory = parameters.getReporterFactory();
 		try {
 			RunListener runListener = reporterFactory.createReporter();
+			ConsoleOutputCapture.startCapture((ConsoleOutputReceiver) runListener);
```
[the remark this answers is not in the document]
Comment (BogdanStirbat) [reply junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/JUnitPlatformProvider.java]: Hi,

Thanks for the review and comments.
Regarding this change, I don't understand where should be added. Should only tests be changed?

**Comment by bedla** [comment]
Hi, I have error test case related to this PR. See https://github.com/bedla/maven-surefire/tree/surefire-2.19.1-junit5 test org.apache.maven.surefire.its.ForkConsoleOutputWithErrorsIT#aaa . Problem is related to communication protocol between surefire and its forked client. Failing test (on my Windows 10 machine) is here forkConsoleOutput.Test2#testXXX . Exception throw can be found on generated log.txt file in /target dir or at the bottom of ForkConsoleOutputWithErrorsIT.java file.
Problem is with string `\n1,5,Foo,Bar,Xxx,Yyy,"{\n` where `1` is command directive. This log output has to be escaped surefire method org.apache.maven.surefire.booter.ForkingRunListener#writeTestOutput do.
Hope it helps you.
Thx,
Ivos

File: junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/JUnitPlatformProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcphilipp) [reply junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/JUnitPlatformProvider.java]: Yes, I was referring to the tests.

**Comment by marcphilipp** [comment]
@BogdanStirbat Do you have time to work on this PR? Since it would also solve #1096 I'd like to see it merged soon. If you don't have time, just let us know and we can pick it up in your stead.

File: junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/JUnitPlatformProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (BogdanStirbat) [reply junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/JUnitPlatformProvider.java]: Hi,

In order to invoke `org.apache.maven.surefire.booter.ProviderFactory.invoke()`, I had to mock a lot of parameters, and I didn't knew all values. 
Also, by calling `org.apache.maven.surefire.booter.ProviderFactory.invoke()` and not `JUnitPlatformProvider.invoke()`, I don't understand how `JUnitPlatformProvider`'s implementation will be tested.

I have a proposal, to restore System.out and System.err, as in method `ProviderFactory.invoke()`. If it's not the right thing to do, please tell.

**Comment by BogdanStirbat** [comment]
Yes, I have time.
Please review the last changes. If something more needs to be added, please let me know.

**Comment by marcphilipp** [comment]
Thanks, @BogdanStirbat! 👍 

I added a test to check the capturing is actually done in 2391334ba2c674c8107baae34b8dc62406042fd7.

**Comment by BogdanStirbat** [comment]
Thanks!

I didn't observed this elegant way of unit testing the behavior.

**Comment by marcphilipp** [comment]
No worries, thanks for your work on the PR!

**Comment by BogdanStirbat** [comment]
No problem. It was my pleasure.

**Comment by bedla** [comment]
I can confirm, that test-case I provided is not failing on current master. Thank you for fix. Ivos

**Comment by marcphilipp** [comment]
Thanks for the confirmation, @bedla!

**Pull request closed without merging**